compiler/rustc_middle/src/ty/query/on_disk_cache.rs

```diff
@@ -765,7 +753,7 @@ impl<'a, 'tcx> Decodable<CacheDecoder<'a, 'tcx>> for &'tcx [Span] {
 //- ENCODING -------------------------------------------------------------------
 
 /// This trait is a hack to work around specialization bug #55243.
-trait OpaqueEncoder: Encoder {
+pub trait OpaqueEncoder: Encoder {
```

Comment:
This trait is a hack. Ideally, all `E: OpaqueEncoder` should be `E = opaque::Encoder`. It needs to be public because it appears in `CacheEncoder`, which is public because it appears in `DepKindTrait`. Removing `E: OpaqueEncoder` from `CacheEncoder` definition breaks specialization (I don't know why).